hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieNativeAvroHFileReader.java

```diff
@@ -244,14 +248,32 @@ private static GenericRecord getRecordFromKeyValue(KeyValue keyValue,
         readerSchema);
   }
 
-  private synchronized HFileReader getSharedHFileReader() {
-    try {
-      if (!sharedHFileReader.isPresent()) {
-        sharedHFileReader = Option.of(newHFileReader());
+  private byte[] getHFileMetaInfoFromCache(String key) throws IOException {
+    if (!PRELOADED_META_INFO_KEYS.contains(key)) {
+      throw new IllegalStateException("HoodieNativeAvroHFileReader#getHFileMetaInfoFromCache"
+          + " should only be called on supported meta info keys; this key is not supported: "
+          + key);
+    }
+    byte[] bytes = metaInfoMap.get(key);
```

Comment:
Can we do a compute if absent here?

Reply:
I think it would be better to check if the map exists at all or not. We might reread a bunch of times if 
```
Option<byte[]> metaInfo = reader.getMetaInfo(new UTF8StringKey(metaInfoKey));
```
is ever option.empty

Reply:
> Can we do a compute if absent here?

I'm updating the whole map so it is anti-pattern to use `computeIfAbsent`.

I revised the checks with a boolean to indicate if the cache is preloaded.